applications/GeoMechanicsApplication/tests/cpp_tests/test_time_loop_executor.cpp

```diff
@@ -0,0 +1,359 @@
+    TimeStepEndState start_state;
+    start_state.convergence_state = TimeStepEndState::ConvergenceState::converged;
+    const auto step_states = executor.Run(start_state);
+
+    KRATOS_EXPECT_EQ(1, step_states.size());
+    KRATOS_EXPECT_TRUE(step_states[0].NonConverged());
+}
+
+KRATOS_TEST_CASE_IN_SUITE(TimeLoopReturnsEndTimesAfterRunningAnAlwaysConvergingSolverStrategy, KratosGeoMechanicsFastSuite)
+{
+    TimeLoopExecutor executor;
+    const auto increments = std::vector<double>{1.1, 1.2, 1.3, 1.4};
+    executor.SetTimeIncrementor(std::make_unique<PrescribedTimeIncrementor>(increments));
+    executor.SetSolverStrategyTimeIncrementor(std::make_unique<DummySolverStrategy>(TimeStepEndState::ConvergenceState::converged));
+    TimeStepEndState start_state;
+    start_state.convergence_state = TimeStepEndState::ConvergenceState::converged;
+    const auto step_states = executor.Run(start_state);
+
+    KRATOS_EXPECT_DOUBLE_EQ(std::accumulate(increments.begin(), increments.end(), 0.), step_states.back().time);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(GetThreeCyclesWhenThreeCyclesAreNeeded, KratosGeoMechanicsFastSuite)
+{
+    TimeLoopExecutor executor;
+    const auto wanted_num_of_cycles_per_step = 3;
+    executor.SetTimeIncrementor(std::make_unique<FixedCyclesTimeIncrementor>(wanted_num_of_cycles_per_step));
+    executor.SetSolverStrategyTimeIncrementor(std::make_unique<DummySolverStrategy>(TimeStepEndState::ConvergenceState::converged));
+    const auto step_states = executor.Run(TimeStepEndState{});
+
+    // The test time incrementor assumes fixed time increments of 0.5 and an end time of 1.0
+    KRATOS_EXPECT_EQ(2, step_states.size());
+    KRATOS_EXPECT_DOUBLE_EQ(0.5, step_states[0].time);
+    KRATOS_EXPECT_EQ(wanted_num_of_cycles_per_step, step_states[0].num_of_cycles);
+    KRATOS_EXPECT_DOUBLE_EQ(1.0, step_states[1].time);
+    KRATOS_EXPECT_EQ(wanted_num_of_cycles_per_step, step_states[1].num_of_cycles);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(GetFiveCyclesWhenFiveCyclesAreNeeded, KratosGeoMechanicsFastSuite)
+{
+    TimeLoopExecutor executor;
+    const auto wanted_num_of_cycles_per_step = 5;
+    executor.SetTimeIncrementor(std::make_unique<FixedCyclesTimeIncrementor>(wanted_num_of_cycles_per_step));
+    executor.SetSolverStrategyTimeIncrementor(std::make_unique<DummySolverStrategy>(TimeStepEndState::ConvergenceState::converged));
+    const auto step_states = executor.Run(TimeStepEndState{});
+
+    // The test time incrementor assumes fixed time increments of 0.5 and an end time of 1.0
+    KRATOS_EXPECT_EQ(2, step_states.size());
+    KRATOS_EXPECT_DOUBLE_EQ(0.5, step_states[0].time);
+    KRATOS_EXPECT_EQ(wanted_num_of_cycles_per_step, step_states[0].num_of_cycles);
+    KRATOS_EXPECT_DOUBLE_EQ(1.0, step_states[1].time);
+    KRATOS_EXPECT_EQ(wanted_num_of_cycles_per_step, step_states[1].num_of_cycles);
+}
```

Comment:
I don't really see why we need an extra test for checking 5 specifically. Is there a reason the 3 would work and the 5 wouldn't?